File: wiki-plugin/scripts/commit.py
"""Write one structured git commit per ingestion/edit.

The commit message is a compounding asset — audit log, "what changed this
week" feed, manager-report source — so it is emitted here, never freehand by
the agent. This docstring is the format's only specification::

    ingest: <source doc title>

    created: wiki/concept/prepared-statements.md
    updated: wiki/concept/db-connection-pooling.md
    superseded: wiki/source/deploy-capistrano.md -> wiki/source/deploy-github-actions.md
    source-date: 2026-03-01

Git is a **hard dependency**: absent ``git``, or a target that isn't a work
tree, raises :class:`GitError` rather than silently skipping — the time model
depends on the history being complete.

A manifest naming a ``raw_source`` is additionally gated on
:func:`chain_of_evidence.check`, raising :class:`CommitGateError` before
anything is staged. This is the hard block; :mod:`ingest` runs the same check
earlier, at plan-validation time, as a courtesy to the agent.

CLI::

    python commit.py --manifest manifest.json   # commits against the resolved vault
"""
from __future__ import annotations

import sys
from dataclasses import dataclass, field
from pathlib import Path

import chain_of_evidence
import wikipage
from vault_git import GitError, VaultGit


class CommitGateError(RuntimeError):
    """Raised when a manifest fails the chain-of-evidence gate.

    Distinct from :class:`GitError`: a rejected manifest is a planning bug,
    a git failure is an environment problem.
    """


@dataclass
class Manifest:
    """The deterministic description of one ingestion/edit's touched files."""

    title: str
    action: str = "ingest"
    created: list[str] = field(default_factory=list)
    updated: list[str] = field(default_factory=list)
    superseded: list[tuple[str, str]] = field(default_factory=list)
    source_date: str | None = None
    #: The raw/ artifact this ingestion is sourced from, if any. Staged
    #: automatically, so the source document always lands in the same commit
    #: as the pages it produced.
    raw_source: str | None = None

    @classmethod
    def from_dict(cls, d: dict) -> "Manifest":
        return cls(
            title=d["title"],
            action=d.get("action", "ingest"),
            created=list(d.get("created", [])),
            updated=list(d.get("updated", [])),
            superseded=[tuple(pair) for pair in d.get("superseded", [])],
            source_date=d.get("source_date"),
            raw_source=d.get("raw_source"),
        )

    def staged_paths(self) -> list[str]:
        """Every path this manifest touches, de-duplicated, in a stable order."""
        paths: list[str] = []
        for page_ref in self.created:
            paths.append(page_ref)
        for page_ref in self.updated:
            paths.append(page_ref)
        for old, new in self.superseded:
            paths.append(old)
            paths.append(new)
        if self.raw_source:
            paths.append(self.raw_source)
        seen: set[str] = set()
        ordered: list[str] = []
        for page_ref in paths:
            if page_ref not in seen:
                seen.add(page_ref)
                ordered.append(page_ref)
        return ordered


def build_message(manifest: Manifest) -> str:
    """Render ``manifest`` to the structured commit message (see module
    docstring for the format). Deterministic."""
    lines = [f"{manifest.action}: {manifest.title}", ""]
    lines += [f"created: {page_ref}" for page_ref in manifest.created]
    lines += [f"updated: {page_ref}" for page_ref in manifest.updated]
    lines += [f"superseded: {old} -> {new}" for old, new in manifest.superseded]
    if manifest.source_date:
        lines.append(f"source-date: {manifest.source_date}")
    return "\n".join(lines) + "\n"


def _check_chain_of_evidence(root: Path, manifest: Manifest) -> None:
    """Gate the commit on :func:`chain_of_evidence.check`.

    No-op when ``manifest.raw_source`` is unset (a synthesis save has no raw
    artifact to demand a stub for). Pages are read from disk — the caller has
    already written them by the time :func:`commit` runs. A staged page
    missing from disk is silently skipped: that's the caller's bug to report,
    not this gate's.
    """
    if manifest.raw_source is None:
        return

    staged: dict[str, wikipage.WikiPage] = {}
    for page_ref in list(manifest.created) + list(manifest.updated):
        page_path = root / page_ref
        if page_path.is_file():
            staged[page_ref] = wikipage.WikiPage(page_path.read_text(encoding="utf-8"))

    errors = chain_of_evidence.check(staged, manifest.raw_source)
    if errors:
        raise CommitGateError("commit gated: " + "; ".join(errors))


def commit(
    vault_root: Path | str,
    manifest: Manifest,
    git: VaultGit | None = None,
) -> str:
    """Stage the manifest's paths and write one structured commit. Returns the SHA.

    ``git`` is injectable for tests (an in-memory :class:`VaultGit` fake); it
    defaults to real git over ``vault_root``. Git stays a hard dependency:
    absent git, or a root that isn't a work tree, raises :class:`GitError`
    from :meth:`VaultGit.ensure_work_tree`.
    """
    root = Path(vault_root)
    git = git or VaultGit(root)
    git.ensure_work_tree()

    paths = manifest.staged_paths()
    _check_chain_of_evidence(root, manifest)
    if paths:
        git.add(*paths)
    return git.commit(build_message(manifest))


def _main(argv=None) -> int:  # pragma: no cover - thin CLI wrapper
    import argparse
    import json

    import vault

    parser = argparse.ArgumentParser(description="Structured git commit per manifest.")
    parser.add_argument("--manifest", required=True, help="path to a manifest JSON file")
    args = parser.parse_args(argv)

    data = json.loads(Path(args.manifest).read_text(encoding="utf-8"))
    manifest = Manifest.from_dict(data)
    root = vault.resolve_vault_root()
    print(commit(root, manifest))
    return 0


if __name__ == "__main__":  # pragma: no cover
    sys.exit(_main())

File: enchiridion-go/internal/commit/commit.go
// Package commit writes one structured git commit per ingestion/edit.
// Ported from `wiki-plugin/scripts/commit.py`.
//
// The commit message is a compounding asset — audit log, "what changed this
// week" feed, manager-report source — so it is emitted here, never freehand
// by the agent. This doc comment is the format's only specification:
//
//	ingest: <source doc title>
//
//	created: wiki/concepts/prepared-statements.md
//	updated: wiki/concepts/db-connection-pooling.md
//	superseded: wiki/sources/deploy-capistrano.md -> wiki/sources/deploy-github-actions.md
//	source-date: 2026-03-01
//
// Git is a **hard dependency**: a root that isn't a work tree is an error,
// never a silent skip — the time model depends on the history being
// complete. The Go port embeds go-git rather than shelling out (ADR-0011),
// so "git is missing from PATH" can no longer arise; what remains is "root
// is not a git work tree".
//
// A manifest naming a RawSource is additionally gated on
// [chainofevidence.Check], failing before anything is staged. This is the
// hard block; the ingest package runs the same check earlier, at
// plan-validation time, as a courtesy to the agent.
package commit

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/dhague/enchiridion/enchiridion-go/internal/chainofevidence"
	"github.com/dhague/enchiridion/enchiridion-go/internal/wikipage"
)

// ErrGate is returned when a manifest fails the chain-of-evidence gate.
//
// Distinct from a git failure: a rejected manifest is a planning bug, a git
// failure is an environment problem.
var ErrGate = errors.New("commit gated")

// Git is the slice of [vaultgit.Repo] this package needs, named as an
// interface so tests can commit against an in-memory fake instead of a real
// repository.
type Git interface {
	IsWorkTree() bool
	Add(paths ...string) error
	Commit(message string) (string, error)
}

// Supersession is one `old -> new` pair in a manifest.
type Supersession struct {
	Old string
	New string
}

// UnmarshalJSON decodes the two-element `[old, new]` array the manifest
// format uses for a supersession pair.
func (s *Supersession) UnmarshalJSON(data []byte) error {
	var pair []string
	if err := json.Unmarshal(data, &pair); err != nil {
		return err
	}
	if len(pair) != 2 {
		return fmt.Errorf("superseded entry must be [old, new], got %d elements", len(pair))
	}
	s.Old, s.New = pair[0], pair[1]
	return nil
}

// MarshalJSON writes the same two-element array [Supersession.UnmarshalJSON]
// reads.
func (s Supersession) MarshalJSON() ([]byte, error) {
	return json.Marshal([]string{s.Old, s.New})
}

// Manifest is the deterministic description of one ingestion/edit's touched
// files.
type Manifest struct {
	Title      string         `json:"title"`
	Action     string         `json:"action"`
	Created    []string       `json:"created"`
	Updated    []string       `json:"updated"`
	Superseded []Supersession `json:"superseded"`
	SourceDate string         `json:"source_date"`
	// RawSource is the raw/ artifact this ingestion is sourced from, if any.
	// Staged automatically, so the source document always lands in the same
	// commit as the pages it produced.
	RawSource string `json:"raw_source"`
}

// defaultAction is the verb a manifest that names none commits under.
const defaultAction = "ingest"

// StagedPaths returns every path this manifest touches, de-duplicated, in a
// stable order.
func (m Manifest) StagedPaths() []string {
	var paths []string
	paths = append(paths, m.Created...)
	paths = append(paths, m.Updated...)
	for _, s := range m.Superseded {
		paths = append(paths, s.Old, s.New)
	}
	if m.RawSource != "" {
		paths = append(paths, m.RawSource)
	}

	seen := map[string]bool{}
	ordered := make([]string, 0, len(paths))
	for _, path := range paths {
		if !seen[path] {
			seen[path] = true
			ordered = append(ordered, path)
		}
	}
	return ordered
}

// BuildMessage renders manifest to the structured commit message (see the
// package comment for the format). Deterministic.
func BuildMessage(m Manifest) string {
	action := m.Action
	if action == "" {
		action = defaultAction
	}
	lines := []string{action + ": " + m.Title, ""}
	for _, pageRef := range m.Created {
		lines = append(lines, "created: "+pageRef)
	}
	for _, pageRef := range m.Updated {
		lines = append(lines, "updated: "+pageRef)
	}
	for _, s := range m.Superseded {
		lines = append(lines, "superseded: "+s.Old+" -> "+s.New)
	}
	if m.SourceDate != "" {
		lines = append(lines, "source-date: "+m.SourceDate)
	}
	return strings.Join(lines, "\n") + "\n"
}

// checkChainOfEvidence gates the commit on [chainofevidence.Check].
//
// A no-op when RawSource is unset (a synthesis save has no raw artifact to
// demand a stub for). Pages are read from disk — the caller has already
// written them by the time [Commit] runs. A staged page missing from disk is
// silently skipped: that's the caller's bug to report, not this gate's.
func checkChainOfEvidence(root string, m Manifest) error {
	if m.RawSource == "" {
		return nil
	}

	staged := map[string]wikipage.Page{}
	for _, pageRef := range append(append([]string{}, m.Created...), m.Updated...) {
		path := filepath.Join(root, filepath.FromSlash(pageRef))
		text, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		staged[pageRef] = wikipage.Page{Text: string(text)}
	}

	problems, err := chainofevidence.Check(staged, m.RawSource)
	if err != nil {
		return err
	}
	if len(problems) > 0 {
		return fmt.Errorf("%w: %s", ErrGate, strings.Join(problems, "; "))
	}
	return nil
}

// Commit stages the manifest's paths and writes one structured commit,
// returning the SHA.
//
// git is injectable for tests; pass a [vaultgit.Repo] over root in
// production. Git stays a hard dependency: a root that isn't a work tree is
// an error, not a skip.
func Commit(root string, m Manifest, git Git) (string, error) {
	if !git.IsWorkTree() {
		return "", fmt.Errorf("%s is not a git work tree; the vault's history is not optional", root)
	}
	if err := checkChainOfEvidence(root, m); err != nil {
		return "", err
	}
	if paths := m.StagedPaths(); len(paths) > 0 {
		if err := git.Add(paths...); err != nil {
			return "", err
		}
	}
	return git.Commit(BuildMessage(m))
}
